notify/templates.go

```diff
@@ -0,0 +1,240 @@
+	Results []TestTemplatesResult      `json:"results"`
+	Errors  []TestTemplatesErrorResult `json:"errors,omitempty"`
+}
+
+type TestTemplatesResult struct {
+	// Name of the associated template definition for this result.
+	Name string `json:"name"`
+
+	// Interpolated value of the template.
+	Text string `json:"text"`
+}
+
+type TestTemplatesErrorResult struct {
+	// Name of the associated template for this error. Will be empty if the Kind is "invalid_template".
+	Name string `json:"name,omitempty"`
+
+	// Kind of template error that occurred.
+	Kind TemplateErrorKind `json:"kind"`
+
+	// Error message.
+	Message string `json:"message"`
+}
+
+type TemplateErrorKind string
+
+const (
+	InvalidTemplate TemplateErrorKind = "invalid_template"
+	ExecutionError  TemplateErrorKind = "execution_error"
+)
+
+// TestTemplate tests the given template string against the given alerts. Existing templates are used to provide context for the test.
+// If an existing template of the same filename as the one being tested is found, it will not be used as context.
+func (am *GrafanaAlertmanager) TestTemplate(ctx context.Context, c TestTemplatesConfigBodyParams) (*TestTemplatesResults, error) {
+	definitions, err := parseTestTemplate(c.Name, c.Template)
+	if err != nil {
+		return &TestTemplatesResults{
+			Errors: []TestTemplatesErrorResult{{
+				Kind:    InvalidTemplate,
+				Message: err.Error(),
+			}},
+		}, nil
+	}
+
+	// Recreate the current template without the definition blocks that are being tested. This is so that any blocks that were removed don't get defined.
+	paths := make([]string, 0)
+	for _, name := range am.templates {
+		if name == c.Name {
+			// Skip the existing template of the same name as we're going to parse the one for testing instead.
+			continue
+		}
+		paths = append(paths, filepath.Join(am.workingDirectory, name))
+	}
+
+	// Parse current templates.
+	var newTextTmpl *tmpltext.Template
+	var captureTemplate template.Option = func(text *tmpltext.Template, _ *tmplhtml.Template) {
+		newTextTmpl = text
+	}
+	newTmpl, err := templates.FromGlobs(paths, captureTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse test template.
+	_, err = newTextTmpl.Parse(c.Template)
+	if err != nil {
+		// This shouldn't happen since we already parsed the template above.
+		return nil, err
+	}
+
+	externalURL, err := url.Parse(am.ExternalURL())
+	if err != nil {
+		return nil, err
+	}
+	newTmpl.ExternalURL = externalURL
+
+	// Prepare the context.
+	alerts := v2.OpenAPIAlertsToAlerts(c.Alerts)
+	ctx = notify.WithReceiverName(ctx, "test receiver")
+	ctx = notify.WithGroupLabels(ctx, model.LabelSet{"group_label": "group_label_value"})
+
+	var tmplErr error
+	templater, _ := templates.TmplText(ctx, newTmpl, alerts, am.logger, &tmplErr)
+
+	// Iterate over each definition in the template and evaluate it.
+	var results TestTemplatesResults
+	for _, def := range definitions {
+		s := fmt.Sprintf(`{{ template "%s" . }}`, def)
```

Comment:
I presume `"` is not allowed in template names, perhaps we should check just in case?

Reply:
If `def` contained a quote (or was otherwise invalid) it would not be parseable and would have failed with `invalid_template` earlier.